File: Iteration_2/src/test/java/TestAdvisor.java
import java.util.ArrayList;
import java.util.Date;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import org.junit.Before;
import org.junit.Test;

public class TestAdvisor {

    private Advisor advisor;
    private Student student;
    private MandatoryCourse course;
    private Transcript transcript;
    ArrayList<Course> courses;
    ArrayList<CourseSection> courseSections;
    
    @Before
    public void setUp() {
        // Mock data setup for testing. This sets up the initial data used by the tests.
        courses = new ArrayList<>();
        course = new MandatoryCourse("IE3107" , "Modeling", 8);
        CourseSection courseSection = new CourseSection("IE3107.1" , 8);
        courseSection.setParentCourse(course);
        courseSections = new ArrayList<>();

        courses.add(course);
        courseSections.add(courseSection);
        course.setCourseSections(courseSections);
        // Creating a new Transcript with empty course lists.
        transcript = new Transcript(new ArrayList<>(), new ArrayList<>(), new ArrayList<>());
        
        // Creating a sample student with basic data.
        student = new Student("Melisa", "Gezer", new Date(), 'F', transcript, "150120101");
        
        // Adding the student to a list.
        ArrayList<Student> students = new ArrayList<>();
        students.add(student);

        // Creating an advisor and assigning the course and student to the advisor.
        advisor = new Advisor("Mert", "Tufan", new Date(), 'M', "369258147", courses, students);
    }

    @Test
    public void testApproveCourse() {
        // Add the course to the student's waited courses.
        student.getTranscript().getWaitedCourses().addAll(courses);

        // Approve the course for the student through the advisor.
        advisor.approveCourse(student, courseSections.get(0));

        // Assert that the course is moved to the student's current courses list.
        assertTrue("Course should be added to current courses.", student.getTranscript().getCurrentCourses().contains(course));
        // Assert that the course is removed from the waited courses list.
        assertFalse("Course should be removed from waited courses.", student.getTranscript().getWaitedCourses().contains(course));
    }

    @Test
    public void testRejectCourse() {
        // Add the course to the student's waited courses.
        student.getTranscript().getWaitedCourses().add(course);

        // Reject the course for the student through the advisor.
        advisor.rejectCourse(student, course);

        // Assert that the course is removed from the waited courses list.
        assertFalse("Course should be removed from waited courses.", student.getTranscript().getWaitedCourses().contains(course));
    }

    @Test
    public void testAddStudent() {
        // Create a new student to add to the advisor's list.
        Student newStudent = new Student("Nilsu", "Konakli", new Date(), 'F', new Transcript(), "150123036");

        // Add the new student to the advisor's list.
        advisor.addStudent(newStudent);

        // Assert that the new student has been added to the advisor's list of students.
        assertTrue("New student should be added to the advisor's student list.", advisor.getStudents().contains(newStudent));
    }

    @Test
    public void testGetStudents() {
        // Retrieve the list of students from the advisor.
        ArrayList<Student> students = advisor.getStudents();
        
        // Ensure the students list is not null.
        assertNotNull("Students list should not be null.", students);
        // Assert that there is exactly one student in the advisor's list.
        assertEquals("Students list should contain one student.", 1, students.size());
    }

    @Test
    public void testCheckSectionConflict() {
        // Existing course section with a specific time slot.
        TimeSlot existingTimeSlot = new TimeSlot("09:30-10:20", "M2Z11");
        CourseSection existingSection = new CourseSection("CSE101.1", 30);
        existingSection.getTimeSlots().add(existingTimeSlot);
        student.getTranscript().addCurrentSection(existingSection);

        // New course section with a non-conflicting time slot.
        TimeSlot newTimeSlot = new TimeSlot("10:30-11:20", "M2Z12");
        CourseSection newSection = new CourseSection("CSE102.1", 30);
        newSection.getTimeSlots().add(newTimeSlot);

        // Test for no conflict.
        boolean resultNoConflict = advisor.checkSectionConflict(student, newSection);
        assertTrue("Sections should not conflict.", resultNoConflict);

        // New course section with a conflicting time slot.
        TimeSlot conflictingTimeSlot = new TimeSlot("09:30-10:20", "M2Z14");
        newSection.getTimeSlots().clear();
        newSection.getTimeSlots().add(conflictingTimeSlot);

        // Test for conflict.
        boolean resultConflict = advisor.checkSectionConflict(student, newSection);
        assertFalse("Sections should conflict.", resultConflict);
    }
}

File: Iteration_3/TestAdvisor.py
import unittest
from datetime import date
from Advisor import Advisor
from Student import Student
from Course import Course
from CourseSection import CourseSection
from Transcript import Transcript
from Lecturer import Lecturer
from TimeSlot import TimeSlot

class TestAdvisor(unittest.TestCase):
    def setUp(self):
        # Course and section setup
        lecturer = Lecturer("Dr.", "Smith", date(1975, 6, 15), "M", "123456789")
        self.course = Course("IE3107", "Modeling", 8)
        self.course_section = CourseSection(8, self.course, lecturer, "IE3107.1")
        self.course.set_course_sections([self.course_section])

        # Transcript and student setup
        self.transcript = Transcript([], [], [])
        self.student = Student("Melisa", "Gezer", date(2000, 1, 1), 'F', self.transcript, "150120101")

        # Advisor setup
        self.advisor = Advisor("Mert", "Tufan", date(1980, 1, 1), 'M', "369258147", [self.course], [self.student])

    def test_add_student(self):
        new_student = Student("Nilsu", "Konakli", date(2002, 5, 20), 'F', Transcript(), "150123036")
        self.advisor.add_student(new_student)
        self.assertIn(new_student, self.advisor.get_students(), "New student was not added to the advisor's student list.")

   # def test_approve_course(self):
    # Öğrencinin transkriptine bekleyen bir kurs ekliyoruz
       # self.student.get_transcript().add_waited_course(self.course)

    # Kursun onaylanmasını sağlıyoruz
       # self.advisor.approve_course(self.student, self.course_section)

    # Kursun geçerli dersler listesine eklenip eklenmediğini kontrol ediyoruz
       # self.assertIn(self.course, self.student.get_transcript().get_current_courses(), "Course was not added to current courses.")
    
    # Bekleyen derslerden çıkarılıp çıkarılmadığını kontrol ediyoruz
       # self.assertNotIn(self.course, self.student.get_transcript().get_waited_courses(), "Course was not removed from waited courses.")
    
    def test_reject_course(self):
        self.student.get_transcript().add_waited_course(self.course)
        self.advisor.reject_course(self.student, self.course_section)
        self.assertNotIn(self.course, self.student.get_transcript().get_waited_courses(), "Course was not removed from waited courses.")

    def test_get_students(self):
        students = self.advisor.get_students()
        self.assertIsNotNone(students, "The student list should not be None.")
        self.assertEqual(len(students), 1, "The advisor should have exactly one student.")

    def test_check_section_conflict(self):
        time_slot1 = TimeSlot("Monday", "09:30-10:20", "M2Z11")
        time_slot2 = TimeSlot("Monday", "10:30-11:20", "M2Z12")
        self.course_section.set_time_slots([time_slot1])

        new_section = CourseSection(8, self.course, self.advisor, "IE3107.2")
        new_section.set_time_slots([time_slot2])
        self.student.get_transcript().add_current_section(self.course_section)

        no_conflict = self.advisor.check_section_conflict(self.student, new_section)
        self.assertTrue(no_conflict, "There should be no time conflict.")

        new_section.set_time_slots([time_slot1])
        conflict = self.advisor.check_section_conflict(self.student, new_section)
        self.assertFalse(conflict, "A time conflict was not detected.")

if __name__ == '__main__':
    unittest.main(verbosity=2)
